arbiter-core/src/middleware/nonce_middleware.rs

```diff
@@ -0,0 +1,206 @@
+//! The `nonce_middleware` module provides a middleware implementation for
+//! managing nonces for Ethereum-like virtual machines. A nonce is a number that
+//! is used only once in a cryptographic communication. In this case, it is used
+//! to ensure that each transaction sent from the address associated with the
+//! middleware is unique and cannot be replayed.
+//!
+//! Main components:
+//! - [`NonceManagerMiddleware`]: The core middleware implementation.
+//! - [`NonceManagerError`]: Error type for the middleware.
+
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+
+use async_trait::async_trait;
+use ethers::{
+    providers::{Middleware, MiddlewareError, PendingTransaction},
+    types::{transaction::eip2718::TypedTransaction, *},
+};
+use thiserror::Error;
+
+#[derive(Debug)]
+/// Middleware used for calculating nonces locally, useful for signing multiple
+/// consecutive transactions without waiting for them to hit the mempool
+pub struct NonceManagerMiddleware<M> {
+    inner: M,
+    init_guard: futures_locks::Mutex<()>,
+    initialized: AtomicBool,
+    nonce: AtomicU64,
+    address: Address,
+}
+
+impl<M> NonceManagerMiddleware<M>
+where
+    M: Middleware,
+{
+    /// Instantiates the nonce manager with a 0 nonce. The `address` should be
+    /// the address which you'll be sending transactions from
+    pub fn new(inner: M, address: Address) -> Self {
```

Comment:
Do we have any better ideas for setting up this API for creating a none managed middleware? 

@kinrezC do you have thoughts?

Reply:
This is the way ether-rs creates theirs.

Reply:
I'm not saying to change this method, I'm saying that giving the user an easy way to create a nonce managed middleware that is like an agent EOA could be a better devex. If this is the only entry point, then we might want to consider something additional.

Reply:
sure, but that seems like a different issue than the one that this feature close.

Reply:
See my original comment. I just asked if there are better ideas for setting up the API.